lib/PersonalityProvider.jsm

```diff
@@ -0,0 +1,197 @@
+ * It does this by looking at the history text content, title, and description.
+ */
+this.PersonalityProvider = class PersonalityProvider {
+  constructor(
+    timeSegments,
+    parameterSets,
+    maxHistoryQueryResults,
+    version,
+    scores,
+    modelKeys) {
+    this.modelKeys = modelKeys;
+    this.timeSegments = timeSegments;
+    this.parameterSets = parameterSets;
+    this.maxHistoryQueryResults = maxHistoryQueryResults;
+    this.version = version;
+    this.scores = scores;
+    this.store = new PersistentCache("personality-provider", true);
+  }
+
+  async init() {
+    this.interestConfig = await this.getRecipe();
+    this.recipeExecutor = await this.generateRecipeExecutor();
+    this.interestVector = await this.store.get("interest-vector");
+
+    // Fetch a new one if none exists or every set update time.
+    if (!this.interestVector ||
+      (Date.now() - this.interestVector.lastUpdate) >= STORE_UPDATE_TIME) {
+      this.interestVector = await this.createInterestVector();
+      this.interestVector.lastUpdate = Date.now();
+      this.store.set("interest-vector", this.interestVector);
+    }
+    this.initialized = true;
+  }
+
+  getRemoteSettings(name) {
+    return RemoteSettings(name).get();
+  }
+
+  getRecipeExecutor(nbTaggers, nmfTaggers) {
+    return new RecipeExecutor(nbTaggers, nmfTaggers);
+  }
+
+  getNaiveBayesTextTagger(model) {
+    return new NaiveBayesTextTagger(model);
+  }
+
+  getNmfTextTagger(model) {
+    return new NmfTextTagger(model);
+  }
+
+  getNewTabUtils() {
+    return NewTabUtils;
+  }
+
+  /**
+   * Returns a Recipe from remote settings to be consumed by a RecipeExecutor.
+   * A Recipe is a set of instructions on how to processes a RecipeExecutor.
+   */
+  async getRecipe() {
+    if (!this.recipe) {
+      this.recipe = await this.getRemoteSettings("personality-provider-recipe");
+    }
+    return this.recipe[0];
+  }
+
+  /**
+   * Returns a Recipe Executor.
+   * A Recipe Executor is a set of actions that can be consumed by a Recipe.
+   * The Recipe determines the order and specifics of which the actions are called.
+   */
+  async generateRecipeExecutor() {
+    let nbTaggers = [];
+    let nmfTaggers = {};
+    const models = await this.getRemoteSettings("personality-provider-models");
+
+    for (let model of models) {
+      if (!model || !this.modelKeys.includes(model.key)) {
+        continue;
+      }
+
+      if (model.data.model_type === "nb") {
+        nbTaggers.push(this.getNaiveBayesTextTagger(model.data));
+      } else if (model.data.model_type === "nmf") {
+        nmfTaggers[model.data.parent_tag] = this.getNmfTextTagger(model.data);
+      }
+    }
+    return this.getRecipeExecutor(nbTaggers, nmfTaggers);
+  }
+
+  /**
+   * Grabs a slice of browse history for building a interest vector
+   */
+  async fetchHistory(columns, beginTimeSecs, endTimeSecs) {
+    let sql = `SELECT *
```

Comment:
A couple of questions for this query:
* Perhaps you only want to fetch certain columns other than `SELECT *` here
* Columns like `description` and `title` could also be `NULL` other than `""`
* Do you want to set a limit on the return set? It's always a good idea to do so to avoid the unexpected behavior
* You can leverage the `params` of executePlaceQuery function for goodies such as specify the column names and set the limit etc, here is an [example](https://searchfox.org/mozilla-central/source/toolkit/modules/NewTabUtils.jsm#1092)

Reply:
@jonathankoren thoughts on this one? I think you're more familiar with this function currently.

Reply:
Fixing the query to explicitly avoid NULL is definable something we should do. 

I wouldn't put a LIMIT on it, because we have no idea what the upper limit of pages should be. Internal polling on our how histories at Pocket showed that the `WHERE description <> ""` has the habit of reducing the total number of URLs in the history somewhere between 75% and 80%.

The reason why we run SELECT * is that we just wanted to make all the columns available. We ran into a problem before we went down the JS route, where not all the columns (specifically `description`) were not available on the browse history API available to extensions. This future proofs us.

Reply:
Yes, settings a very high LIMIT is definitely better than nothing. I believe you can find a good tradeoff from the model's minimal requirement and the computing limitation. 

> The reason why we run SELECT * is that we just wanted to make all the columns available.

Do you really need to make all the columns available? Explicitly specifying the needed columns has at least two benefits:
* It significantly reduces the IO for the big query results, there is no point to load columns without using them
* It makes the code self-described about what Places columns are in use here